### ISSUE TITLE
[BUG] java.lang.IllegalArgumentException: Accessing record m$2 of type Record["bal" -> Int, "type" -> Const] with the field "acc". Type inference should have caught this.

### ISSUE THREAD

**Issue by vitorenesduarte** [informalsystems/apalache#628]
I started from the `Paxos.tla` in this repository: https://github.com/informalsystems/apalache/blob/unstable/test/tla/Paxos.tla
Then did a small change in the existing annotations (https://github.com/vitorenesduarte/expi/commit/867199e534a4261fc586b3d163f5e690f7e2072f) which resulted in the error below.

**To Reproduce**
```
git clone https://github.com/vitorenesduarte/expi -b apalache-error
apalache-mc check Paxos.tla
```

```
# APALACHE version 0.11.0 build v0.10.0-30-g8c04870
#
# WARNING: This tool is in the experimental stage.
#          Please report bugs at: [https://github.com/informalsystems/apalache/issues]
#
# Usage statistics is OFF. We care about your privacy.
# If you want to help our project, consider enabling statistics with config --enable-stats=true.

Checker options: filename=Paxos.tla, init=, next=, inv=           I@19:55:05.869
Tuning:                                                           I@19:55:06.059
PASS #0: SanyParser                                               I@19:55:06.059
Parsing file /home/vitor.enes/a/expi/Paxos.tla
Parsing file /tmp/Integers.tla
Parsing file /tmp/Naturals.tla
PASS #1: ConfigurationPass                                        I@19:55:06.386
  > Paxos.cfg: Loading TLC configuration                          I@19:55:06.387
  > No TLC configuration found. Skipping.                         I@19:55:06.392
  > Command line option --init is not set. Using Init             I@19:55:06.392
  > Command line option --next is not set. Using Next             I@19:55:06.392
  > Set the initialization predicate to Init                      I@19:55:06.392
  > Set the transition predicate to Next                          I@19:55:06.393
PASS #2: DesugarerPass                                            I@19:55:06.419
  > Desugaring...                                                 I@19:55:06.419
PASS #3: UnrollPass                                               I@19:55:06.445
  > Unroller                                                      I@19:55:06.478
PASS #4: PrimingPass                                              I@19:55:06.495
  > Introducing InitPrimed for Init'                              I@19:55:06.497
PASS #5: CoverAnalysisPass                                        I@19:55:06.515
PASS #6: InlinePass                                               I@19:55:06.516
  > InlinerOfUserOper                                             I@19:55:06.520
  > LetInExpander                                                 I@19:55:06.535
  > InlinerOfUserOper                                             I@19:55:06.540
Leaving only relevant operators: CInitPrimed, Init, InitPrimed, Next I@19:55:06.542
PASS #7: VCGen                                                    I@19:55:06.549
  > No invariant given. Only deadlocks will be checked            I@19:55:06.549
PASS #8: PreprocessingPass                                        I@19:55:06.556
  > Before preprocessing: unique renaming                         I@19:55:06.556
 > Applying standard transformations:                             I@19:55:06.559
  > PrimePropagation                                              I@19:55:06.560
  > Desugarer                                                     I@19:55:06.567
  > UniqueRenamer                                                 I@19:55:06.575
  > Normalizer                                                    I@19:55:06.587
  > Keramelizer                                                   I@19:55:06.596
  > After preprocessing: UniqueRenamer                            I@19:55:06.608
PASS #9: TransitionFinderPass                                     I@19:55:06.620
  > Found 1 initializing transitions                              I@19:55:06.641
  > Found 4 transitions                                           I@19:55:06.654
  > No constant initializer                                       I@19:55:06.654
  > Applying unique renaming                                      I@19:55:06.655
PASS #10: OptimizationPass                                        I@19:55:06.667
 > Applying optimizations:                                        I@19:55:06.669
  > ConstSimplifier                                               I@19:55:06.670
  > ExprOptimizer                                                 I@19:55:06.673
  > ConstSimplifier                                               I@19:55:06.675
PASS #11: AnalysisPass                                            I@19:55:06.681
 > Marking skolemizable existentials and sets to be expanded...   I@19:55:06.682
  > SkolemizationMarker                                           I@19:55:06.682
  > ExpansionMarker                                               I@19:55:06.683
 > Running analyzers...                                           I@19:55:06.686
  > Introduced expression grades                                  I@19:55:06.692
  > Introduced 1 formula hints                                    I@19:55:06.692
PASS #12: BoundedChecker                                          I@19:55:06.692
Step 0: picking a transition out of 1 transition(s)               I@19:55:06.916
Step 1: Transition #1 is disabled                                 I@19:55:06.960
Step 1: picking a transition out of 1 transition(s)               I@19:55:06.969
(Please report an issue at: [https://github.com/informalsystems/apalache/issues],java.lang.IllegalArgumentException: Accessing record m$2 of type Record["bal" -> Int, "type" -> Const] with the field "acc". Type inference should have caught this.)
Unhandled exception                                               E@19:55:06.983
java.lang.IllegalArgumentException: Accessing record m$2 of type Record["bal" -> Int, "type" -> Const] with the field "acc". Type inference should have caught this.
	at at.forsyte.apalache.tla.bmcmt.rules.FunAppRule.applyRecord(FunAppRule.scala:73)
	at at.forsyte.apalache.tla.bmcmt.rules.FunAppRule.apply(FunAppRule.scala:42)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:334)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:365)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:396)
	at at.forsyte.apalache.tla.bmcmt.rules.SetInRule.apply(SetInRule.scala:43)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:334)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:365)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:396)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.mapArg$1(AndRule.scala:74)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.$anonfun$apply$2(AndRule.scala:78)
	at scala.collection.immutable.List.map(List.scala:297)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.apply(AndRule.scala:78)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:334)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:365)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:396)
	at at.forsyte.apalache.tla.bmcmt.rules.SetFilterRule.eachElem$1(SetFilterRule.scala:41)
	at at.forsyte.apalache.tla.bmcmt.rules.SetFilterRule.$anonfun$apply$1(SetFilterRule.scala:47)
	at scala.collection.immutable.List.map(List.scala:293)
	at at.forsyte.apalache.tla.bmcmt.rules.SetFilterRule.apply(SetFilterRule.scala:47)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:334)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:365)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:396)
	at at.forsyte.apalache.tla.bmcmt.rules.LetInRule.bindOperator(LetInRule.scala:44)
	at at.forsyte.apalache.tla.bmcmt.rules.LetInRule.$anonfun$apply$1(LetInRule.scala:25)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:91)
	at at.forsyte.apalache.tla.bmcmt.rules.LetInRule.apply(LetInRule.scala:25)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:334)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:365)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:396)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.skolemExistsInStaticallyNonEmptySet(QuantRule.scala:235)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.skolemExistsInSet(QuantRule.scala:209)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.apply(QuantRule.scala:50)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:334)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:365)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:396)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.mapArg$1(AndRule.scala:74)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.$anonfun$apply$2(AndRule.scala:78)
	at scala.collection.immutable.List.map(List.scala:297)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.apply(AndRule.scala:78)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:334)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:365)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:396)
	at at.forsyte.apalache.tla.bmcmt.rules.LetInRule.bindOperator(LetInRule.scala:44)
	at at.forsyte.apalache.tla.bmcmt.rules.LetInRule.$anonfun$apply$1(LetInRule.scala:25)
	at scala.collection.IndexedSeqOptimized.foldLeft(IndexedSeqOptimized.scala:60)
	at scala.collection.IndexedSeqOptimized.foldLeft$(IndexedSeqOptimized.scala:68)
	at scala.collection.mutable.ArrayBuffer.foldLeft(ArrayBuffer.scala:49)
	at at.forsyte.apalache.tla.bmcmt.rules.LetInRule.apply(LetInRule.scala:25)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:334)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:365)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:396)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.skolemExistsInStaticallyNonEmptySet(QuantRule.scala:235)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.skolemExistsInSet(QuantRule.scala:209)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.apply(QuantRule.scala:50)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:334)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:365)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:396)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.skolemExistsInRange(QuantRule.scala:198)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.apply(QuantRule.scala:42)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:334)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:365)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:396)
	at at.forsyte.apalache.tla.bmcmt.trex.TransitionExecutorImpl.prepareTransition(TransitionExecutorImpl.scala:102)
	at at.forsyte.apalache.tla.bmcmt.trex.FilteredTransitionExecutor.prepareTransition(FilteredTransitionExecutor.scala:34)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.$anonfun$prepareTransitionsAndCheckInvariants$5(SeqModelChecker.scala:112)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.$anonfun$prepareTransitionsAndCheckInvariants$5$adapted(SeqModelChecker.scala:106)
	at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:924)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:923)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.prepareTransitionsAndCheckInvariants(SeqModelChecker.scala:106)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.makeStep(SeqModelChecker.scala:51)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.run(SeqModelChecker.scala:39)
	at at.forsyte.apalache.tla.bmcmt.passes.BoundedCheckerPassImpl.runIncrementalChecker(BoundedCheckerPassImpl.scala:108)
	at at.forsyte.apalache.tla.bmcmt.passes.BoundedCheckerPassImpl.execute(BoundedCheckerPassImpl.scala:83)
	at at.forsyte.apalache.infra.passes.PassChainExecutor.exec$1(PassChainExecutor.scala:22)
	at at.forsyte.apalache.infra.passes.PassChainExecutor.run(PassChainExecutor.scala:37)
	at at.forsyte.apalache.tla.Tool$.runCheck(Tool.scala:173)
	at at.forsyte.apalache.tla.Tool$.$anonfun$run$2(Tool.scala:95)
	at at.forsyte.apalache.tla.Tool$.$anonfun$run$2$adapted(Tool.scala:95)
	at at.forsyte.apalache.tla.Tool$.handleExceptions(Tool.scala:258)
	at at.forsyte.apalache.tla.Tool$.run(Tool.scala:95)
	at at.forsyte.apalache.tla.Tool$.main(Tool.scala:45)
	at at.forsyte.apalache.tla.Tool.main(Tool.scala)
It took me 0 days  0 hours  0 min  1 sec                          I@19:55:06.983
Total time: 1.116 sec                                             I@19:55:06.983
EXITCODE: ERROR (99)
```

**Desktop:**
 - Debian GNU/Linux 10
- openjdk 11.0.9.1 2020-11-04
- APALACHE version 0.11.0 build v0.10.0-30-g8c04870

**Comment by konnov** [comment]
@vitorenesduarte this bug report concerns the old type annotations, which are not used in the new version 0.15.0. Do you like to try it with 0.15.0 and check, whether the bug persists?

**Comment by konnov** [comment]
Closing this one, as we will fix records in #401.
